Allow enabling/disabling filters

adamint opened the pull request:
## Description

In the telemetry filter menu, adds checkboxes for each filter like in console logs settings. Only enabled filters are counted in the badge. Enable/disable state is persisted in the the serialized filter.
![image](https://github.com/user-attachments/assets/0a3435e7-4fd8-4288-aa2c-be3a34f6d3c2)

Used a pause/resume button for the filter dialog (other icon suggestions welcome)
![image](https://github.com/user-attachments/assets/c7a1b779-3840-431f-a962-5018d538fde5)


Fixes #4143

## Checklist

- Is this feature complete?
  - [X] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [ ] Yes
  - [X] No (partially covered by existing tests)
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [X] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [X] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Is this introducing a breaking change?
      - [ ] Yes
        - Link to aspire-docs issue (please use this [`breaking-change` template](https://github.com/dotnet/docs-aspire/issues/new?template=04-breaking-change.yml)):
      - [ ] No
        - Link to aspire-docs issue (please use this [`doc-idea` template](https://github.com/dotnet/docs-aspire/issues/new?template=02-docs-request.yml)):
  - [X] No

adamint committed:
Allow enabling/disabling filters

adamint committed:
Update test

JamesNK commented on src/Aspire.Dashboard/Extensions/TelemetryFilterFormatter.cs:
```diff
@@ -23,7 +23,9 @@ private static string SerializeFilterToString(TelemetryFilter filter)
-        return $"{Escape(filter.Field)}:{condition}:{Escape(filter.Value)}";
+        var state = filter.Enabled ? "enabled" : "disabled";
+
+        return $"{Escape(filter.Field)}:{condition}:{Escape(filter.Value)}:{state}";
```
Could you make it so the `:disabled` is only added when disabled? Most filters will always be enabled. Also, keeps compatibility with bookmarks of existing filters

JamesNK commented on src/Aspire.Dashboard/Extensions/TelemetryFilterFormatter.cs:
```diff
@@ -34,7 +36,7 @@ public static string SerializeFiltersToString(IEnumerable<TelemetryFilter> filte
     {
         var parts = filterString.Split(':');
-        if (parts.Length != 3)
+        if (parts.Length != 4)
```
If we make `:disabled` optional then we'll need to support 3 or 4.

JamesNK commented:
UI feedback:

The tick/empty box icons aren't appropriate here because you can't click on the menu item to toggle. What about play and pause as icons to indicate status instead?

Do you think a pause all/enable all option in the menu would be useful?

JamesNK commented:
The button position look a little sloppy. What do you think of this:

![image](https://github.com/user-attachments/assets/43fb3a5a-8d5e-472f-8ab4-f3bd2308da70)

adamint committed:
do not add :enabled

adamint commented:
> UI feedback:
> 
> The tick/empty box icons aren't appropriate here because you can't click on the menu item to toggle. What about play and pause as icons to indicate status instead? Edit: Instead of play icon, how about the same icon as running from the resources page (although standard purple instead of green). ![image](https://private-user-images.githubusercontent.com/303201/427359319-d2dc5c88-cfc2-45e5-bab9-394810be3b69.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMwOTk2NjAsIm5iZiI6MTc0MzA5OTM2MCwicGF0aCI6Ii8zMDMyMDEvNDI3MzU5MzE5LWQyZGM1Yzg4LWNmYzItNDVlNS1iYWI5LTM5NDgxMGJlM2I2OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyN1QxODE2MDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yYTk4YmIxYjlhMjRiYjNlNzc2MzM4NzRhNmQ3OWYxNjMyZjI1ODE5ZGY1MDI0NGE0YjllYjA0ZTNlZTlhNTI2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.2daU16CJVRFXwW4Z_MLSWNScpm6u6YIrIggpAWgqNuo)
> 
> Do you think a pause all/enable all option in the menu would be useful?

Changed the icons. I do like the pause all/enable all and added that.

![image](https://github.com/user-attachments/assets/cd16229a-16af-4d72-a295-324ae017a487)

adamint commented:
> The button position look a little sloppy. What do you think of this:
> 
> ![image](https://private-user-images.githubusercontent.com/303201/427353863-43fb3a5a-8d5e-472f-8ab4-f3bd2308da70.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMwOTk2NjAsIm5iZiI6MTc0MzA5OTM2MCwicGF0aCI6Ii8zMDMyMDEvNDI3MzUzODYzLTQzZmIzYTVhLThkNWUtNDcyZi04YWI0LWYzYmQyMzA4ZGE3MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyN1QxODE2MDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mNjUwOTcyMDJiMWI1MjQxOWQyMmYyZmQyNDMzMGYzNmJlMDVlYWI3MTA1NWIwYWNkYzM0MThiYTFlOGJlNjlhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.-nuV-oX2mFEFsKzXXCvaW5qmv9SRzqFr-PpCUEkwRFQ)

![image](https://github.com/user-attachments/assets/1b307297-c9d5-47ca-be4e-fff743352e27)

Changed.

adamint committed:
extract get filter items in structured logs/traces to helper method, change icons, move buttons, add disable/enable all

JamesNK commented on src/Aspire.Dashboard/Extensions/TelemetryFilterFormatter.cs:
```diff
@@ -23,7 +23,13 @@ private static string SerializeFilterToString(TelemetryFilter filter)
+        var filterString = $"{Escape(filter.Field)}:{condition}:{Escape(filter.Value)}";
+        if (!filter.Enabled)
+        {
+            filterString += ":disabled";
```
nit: move "disabled" to a const. Use and here in the detection check
```suggestion
            filterString += $":{DisabledText}";
```

JamesNK commented on src/Aspire.Dashboard/Extensions/TelemetryFilterFormatter.cs:
```diff
@@ -61,11 +67,14 @@ public static string SerializeFiltersToString(IEnumerable<TelemetryFilter> filte
 
         var value = Unescape(parts[2]);
 
+        var enabled = parts is not [_, _, _, "disabled"];
```
```suggestion
        var enabled = parts is not [_, _, _, DisabledText];
```

JamesNK commented on src/Aspire.Dashboard/Components/Pages/StructuredLogs.razor:
```diff
@@ -76,7 +76,7 @@
             {
                 <div slot="end">
-                    <FluentCounterBadge HorizontalPosition="70" Count="@(ViewModel.Filters.Count)" Appearance="Appearance.Accent">
+                    <FluentCounterBadge HorizontalPosition="70" Count="@(ViewModel.Filters.Count(filter => filter.Enabled))" Appearance="Appearance.Accent">
```
There are a lot of `filter => filter.Enabled` in this code. What about adding `EnabledFilters` property to the view model, e.g.
```cs
public IEnumerable<...> GetEnabledFilters() => Filters.Count(filter => filter.Enabled);
```

adamint replied:
We would need to add this to the like 5 view models that contain a filter list. How about this instead?

```csharp
public static IEnumerable<TelemetryFilter> GetEnabledFilters(this IEnumerable<TelemetryFilter> filters)
{
  return filters.Where(filter => filter.Enabled);
}
```

JamesNK replied:
Sure

JamesNK commented on src/Aspire.Dashboard/Utils/FilterHelpers.cs:
```diff
@@ -0,0 +1,91 @@
+            filterMenuItems.Add(new MenuButtonItem
+            {
+                Text = dialogsLoc[nameof(Dialogs.FilterDialogEnableAll)],
+                Icon = new Size16.CheckmarkCircle(),
```
```suggestion
                Icon = new Size16.Play(),
```

JamesNK commented on src/Aspire.Dashboard/Utils/FilterHelpers.cs:
```diff
@@ -0,0 +1,91 @@
+using Aspire.Dashboard.Model.Otlp;
+using Aspire.Dashboard.Resources;
+using Microsoft.Extensions.Localization;
+using Microsoft.FluentUI.AspNetCore.Components.Icons.Regular;
```
nit: This is how we reference icons everywhere else.
```suggestion
using Icons = Microsoft.FluentUI.AspNetCore.Components.Icons;
```

JamesNK commented:
Looks good. Some clean up feedback.

adamint committed:
Merge branch 'main' into dev/adamint/allow-toggling-traces-structured-log-filters

adamint committed:
clean up
